[SPARK-17031][SQL] Add `Scanner` operator to wrap the optimized plan directly in planner

jiangxb1987 opened the pull request:
## What changes were proposed in this pull request?

Added `Scanner` operator to wrap the optimized plan directly in planner, it holds project lists as well as filter predicates.
Updated relative `Analyzer` and `Optimizer` rules.
## How was this patch tested?

Existing testcases.

jiangxb1987 committed:
insert a Scanner operator over relation and combine Project/Filter(s) with it.

jiangxb1987 committed:
resolve testcases fail.

jiangxb1987 committed:
fix testcases fail.

jiangxb1987 committed:
fix Scalastyle check fails.

SparkQA commented:
**[Test build #63679 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63679/consoleFull)** for PR 14619 at commit [`b1e224c`](https://github.com/apache/spark/commit/b1e224c148280ad1e420b2d9b1dbb7c49f32f931).

SparkQA commented:
**[Test build #63679 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63679/consoleFull)** for PR 14619 at commit [`b1e224c`](https://github.com/apache/spark/commit/b1e224c148280ad1e420b2d9b1dbb7c49f32f931).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63679/
Test FAILed.

hvanhovell commented:
Could you elaborate on what you are trying to do here? I am missing context here. What is the advantage of doing this?

jiangxb1987 committed:
update failed hive tests.

SparkQA commented:
**[Test build #63696 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63696/consoleFull)** for PR 14619 at commit [`0fc7d00`](https://github.com/apache/spark/commit/0fc7d0060a7be633f2a23954b471b71821249882).

jiangxb1987 commented:
@hvanhovell This idea is inspired by @cloud-fan, as he stated in [comment](https://github.com/apache/spark/pull/13893#discussion_r73305098), we'd better have a wrapper node for scan, so that the planner may match the wrapper node directly instead of resolving the whole plan using `PhysicalOperator`.

SparkQA commented:
**[Test build #63696 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63696/consoleFull)** for PR 14619 at commit [`0fc7d00`](https://github.com/apache/spark/commit/0fc7d0060a7be633f2a23954b471b71821249882).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63696/
Test PASSed.

rxin commented:
I'm pretty confused by this as well. Is this just collapsing Filter, Project, and an arbitrary node into a single logical node?

cloud-fan commented:
see discussion here: https://github.com/apache/spark/pull/13893#discussion_r73117855

Currently we collect the projects and filters on scan node at planner by `PhysicalOperator.unapply`. The `PhysicalOperator.unapply` mostly duplicates the logic from column pruning and filter push down rules in optimizer, but doesn't handle non-deterministic expressions well. By adding a wrapper node on scan, we can push down projects and filters to scan node at optimizer phase, and reuse the existing rules. Thus we can eliminate the duplicated codes and handle non-deterministic expressions.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1636,6 +1638,30 @@ object ReplaceExceptWithAntiJoin extends Rule[LogicalPlan] {
+ * Insert a [[Scanner]] operator over [[MultiInstanceRelation]], then combine [[Project]]s and
+ * [[Filter]]s with this operator, then the planner can match on the [[Scanner]] node directly.
+ */
+object InsertRelationScanner extends Rule[LogicalPlan] with PredicateHelper {
```
This doesn't reuse the existing `ColumnPruning` and `PushDownPredicate` rules. I'd like to add the wrapper at the very beginning, e.g. `SessionCatalog.lookUpRelation`.

jiangxb1987 replied:
Roger that, will change it and retest.

jiangxb1987 committed:
move InsertRelationScanner operator to the early stage of analyzing.

jiangxb1987 committed:
refactor CovertToLocalRelation rule.

jiangxb1987 committed:
fix failed test cases.

jiangxb1987 committed:
fix bug in PushDownPredicate.

jiangxb1987 committed:
fix ConvertToLocalRelation rule.

jiangxb1987 committed:
bugfix.

jiangxb1987 committed:
bugfix.

jiangxb1987 committed:
bugfix

jiangxb1987 committed:
fix SQLBuilder fails.

jiangxb1987 committed:
bugfix.

jiangxb1987 committed:
bugfix.

jiangxb1987 committed:
fix scalastyle fail.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64532/
Test FAILed.

SparkQA commented:
**[Test build #64533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64533/consoleFull)** for PR 14619 at commit [`6eb5bd7`](https://github.com/apache/spark/commit/6eb5bd76b833c3bd21bc47943187e9acb414d098).

SparkQA commented:
**[Test build #64533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64533/consoleFull)** for PR 14619 at commit [`6eb5bd7`](https://github.com/apache/spark/commit/6eb5bd76b833c3bd21bc47943187e9acb414d098).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64533/
Test FAILed.

jiangxb1987 committed:
update SubqueryAlias references.

SparkQA commented:
**[Test build #64544 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64544/consoleFull)** for PR 14619 at commit [`1d502b6`](https://github.com/apache/spark/commit/1d502b683d3183e9f30deca6939b075d09418954).

SparkQA commented:
**[Test build #64544 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64544/consoleFull)** for PR 14619 at commit [`1d502b6`](https://github.com/apache/spark/commit/1d502b683d3183e9f30deca6939b075d09418954).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64544/
Test FAILed.

jiangxb1987 commented:
@cloud-fan I've moved the `InsertRelationScanner` rule to `Analyzer`, after relations and expressions are resolved. To reuse analyze and optimize rules, I updated relative rules such as `CleanupAliases`、 `ColumnPruning`、 `PushDownPredicate`、 `InferFiltersFromConstraints`、 `ConvertToLocalRelation`、 `PropagateEmptyRelation`, I also added new rules to combine and prune `Scanner` operators. Besides, I made some change in subquery related rules and recently found they have been refactored.
Now that only a few of test cases is still failing, which should be easy to fix. But, I realized adding a wrapper node over every relation maybe not a idea that is perfect enough for the following reasons:
Firstly, scan a relation is not among basic operators in SQL language, when we declare a relation, we imply it should be scanned, so It seems semantically duplicate to declare a `Scanner` node over a relation or calling `relation.scanner()`. Besides, to add this wrapper node, we have to make a new assumption that no other operators should be inserted between `Scanner` and its corresponding relation, this brought in more complexity.
Secondly, a wrapper node should contain the output, predicates that can be used in partition pruning, and a relation to be scanned. But this may cause complex situation in some cases, for example, in `InferFiltersFromConstraints`, we have to covert expression in filters to alias name when we collect valid constraints, because output maybe alias and filters have to use child expression, this behavor is not needed in other operators.
At last, I feel adding such a operator have caused too many changes, perhaps we should make some improvement on `PhysicalOperation`, until we figure out a way comprehensively better than current method.

After all, I'm passionate to this improvement and will try my best to contribute, please correct me if I'm wrong, thank you!

jiangxb1987 committed:
bugfix

jiangxb1987 committed:
update PredicateSubquery optimize rule.

SparkQA commented:
**[Test build #65033 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65033/consoleFull)** for PR 14619 at commit [`957d784`](https://github.com/apache/spark/commit/957d784de7c1f56230dfca619eb6895017984224).

SparkQA commented:
**[Test build #65033 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65033/consoleFull)** for PR 14619 at commit [`957d784`](https://github.com/apache/spark/commit/957d784de7c1f56230dfca619eb6895017984224).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65033/
Test FAILed.

jiangxb1987 committed:
fix scalastyle fail.

SparkQA commented:
**[Test build #65044 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65044/consoleFull)** for PR 14619 at commit [`230d65d`](https://github.com/apache/spark/commit/230d65d93394e0129f3ab5a268f6f20ae5696876).

SparkQA commented:
**[Test build #65044 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65044/consoleFull)** for PR 14619 at commit [`230d65d`](https://github.com/apache/spark/commit/230d65d93394e0129f3ab5a268f6f20ae5696876).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65044/
Test FAILed.
